Issue#628 NPE in CTE

stumc opened the pull request:
I found a few problems:
- duplicate table view not cleared down in CTE creation
- wrong order of command container stop cleanup calls (which I reverted)
- column name bug - causing NPE in meta data query accessing temporary's column name when querySQL is null
- two check assertions  coming out of meta-lock (see comments)
- ridiculous column names generated with non-printing characters, CR/LF with no standards or checks which exist in our emulated DBs

stumc committed:
First round of investigations, column namer added, noticed duplicate tableView not being cleaned down correctly

stumc committed:
First round of investigations, column namer added, noticed duplicate tableView not being cleaned down correctly

stumc committed:
Removing CR/LF from column names

stumc committed:
Removing CR/LF from column names

stumc committed:
Added standard column namer

stumc committed:
Added standard column namer

stumc committed:
Standardized names, removed sysout comments from test

stumc committed:
Standardized names, removed sysout comments from test

stumc committed:
Asserting on column names

stumc committed:
Asserting on column names

grandinj commented on h2/src/test/org/h2/test/db/TestGeneralCommonTableQueries.java:
```diff
@@ -469,4 +469,82 @@ private void testColumnNames() throws Exception {
+            +"FROM A                                   \n"
+            +"GROUP BY A.val";
+
+        for(@SuppressWarnings("unused") int queryRunTries: new int[]{1,2,3}){
```
this seems a little clever compared to a normal for loop :-)

stumc replied:
Reverted

grandinj commented on h2/src/main/org/h2/util/ColumnNamer.java:
```diff
@@ -0,0 +1,81 @@
+                columnName = null;
+            }
+        }
+        // try a name derived form the column expression SQL
```
form->from

grandinj commented on h2/src/main/org/h2/util/ColumnNamer.java:
```diff
@@ -0,0 +1,81 @@
+        if (columnNameOverride != null){
+            columnName = columnNameOverride;
+        }  
+        // try a name form the column alias
```
form->from

grandinj commented on h2/src/main/org/h2/util/ColumnNamer.java:
```diff
@@ -0,0 +1,81 @@
+    } 
+    
+    /**
+     * Create a standardized column name that isn't null or and doesn't have a CR/LF in it.
```
"or and" ?

grandinj commented on h2/src/main/org/h2/util/ColumnNamer.java:
```diff
@@ -0,0 +1,81 @@
+        return getColumnName(expr,indexOfColumn,(String) null);
+    }
+    /**
+     * Create a standardized column name that isn't null or and doesn't have a CR/LF in it.
```
"or and" ?

grandinj commented on h2/src/main/org/h2/util/ColumnNamer.java:
```diff
@@ -0,0 +1,81 @@
+public class ColumnNamer {
+ 
+    /**
+     * Create a standardized column name that isn't null or and doesn't have a CR/LF in it.
```
"or and" ?

grandinj reviewed (changes requested):
Lots of errors in the unit tests?

grandinj commented:
I have also just pushed a fix to do with column names on VIEWs, which might (but probably wont) affect this change, so you might want to rebase

stumc committed:
Tried another getSQL() strategy for column naming

stumc committed:
Finnicky situations and desperate measures

stumc committed:
Merged back with branch

stumc committed:
Fixed too smart by far for-loop

stumc commented:
uggghhh - I think my rebase brought in all the MERGE USING change too into the review...

grandinj commented:
yeah, can you merge this with master or rebase, either will reduce it back down to a reviewable size :-)

stumc committed:
Merge branch 'master' into Issue#628

grandinj commented on h2/src/main/org/h2/engine/Session.java:
```diff
@@ -373,6 +373,8 @@ public void addLocalTempTable(Table table) {
      */
     public void removeLocalTempTable(Table table) {
+        // Exception thrown in org.h2.engine.Database.removeMeta if line below is missing with TestGeneralCommonTableQueries
+        database.lockMeta(this);
```
I;m not seeing these exceptions when running that test - how do you reproduce this. Not keen to scatter more locking around unnecessarily

stumc replied:
You have to comment out the database.lockMeta(this); line shown and run the TestGeneralCommonTableQueries test:

Exception in thread "main" org.h2.jdbc.JdbcSQLException: General error: "java.lang.RuntimeException: Unexpected code path" [50000-196]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:168)
	at org.h2.message.DbException.convert(DbException.java:295)
	at org.h2.message.DbException.toSQLException(DbException.java:268)
	at org.h2.message.TraceObject.logAndConvert(TraceObject.java:358)
	at org.h2.jdbc.JdbcPreparedStatement.execute(JdbcPreparedStatement.java:220)
	at org.h2.test.db.TestGeneralCommonTableQueries.testCreateTable(TestGeneralCommonTableQueries.java:378)
	at org.h2.test.db.TestGeneralCommonTableQueries.test(TestGeneralCommonTableQueries.java:42)
	at org.h2.test.db.TestGeneralCommonTableQueries.main(TestGeneralCommonTableQueries.java:26)
Caused by: java.lang.RuntimeException: Unexpected code path
	at org.h2.message.DbException.throwInternalError(DbException.java:242)
	at org.h2.message.DbException.throwInternalError(DbException.java:255)
	at org.h2.engine.Database.removeMeta(Database.java:945)
	at org.h2.table.Table.removeChildrenAndResources(Table.java:553)
	at org.h2.table.TableView.removeChildrenAndResources(TableView.java:422)
	at org.h2.engine.Session.removeLocalTempTable(Session.java:381)
	at org.h2.command.CommandContainer.stop(CommandContainer.java:130)
	at org.h2.command.Command.executeUpdate(Command.java:290)
	at org.h2.jdbc.JdbcPreparedStatement.execute(JdbcPreparedStatement.java:207)
	... 3 more

stumc committed:
Merge remote-tracking branch 'origin/master' into Issue#628

stumc commented on h2/src/main/org/h2/engine/Session.java:
```diff
@@ -974,6 +976,8 @@ private void cleanTempTables(boolean closeSession) {
                         table.setModified();
                         it.remove();
+                        // Exception thrown in org.h2.engine.Database.removeMeta if line below is missing with TestDeadlock
+                        database.lockMeta(this);
```
Also commented out this "fix" line and got:
Exception in thread "main" org.h2.jdbc.JdbcSQLException: General error: "java.lang.RuntimeException: Unexpected code path" [50000-196]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:168)
	at org.h2.message.DbException.convert(DbException.java:295)
	at org.h2.message.DbException.toSQLException(DbException.java:268)
	at org.h2.message.TraceObject.logAndConvert(TraceObject.java:358)
	at org.h2.jdbc.JdbcConnection.close(JdbcConnection.java:413)
	at org.h2.test.db.TestDeadlock.testTemporaryTablesAndMetaDataLocking(TestDeadlock.java:413)
	at org.h2.test.db.TestDeadlock.test(TestDeadlock.java:52)
	at org.h2.test.db.TestDeadlock.main(TestDeadlock.java:46)
Caused by: java.lang.RuntimeException: Unexpected code path
	at org.h2.message.DbException.throwInternalError(DbException.java:242)
	at org.h2.message.DbException.throwInternalError(DbException.java:255)
	at org.h2.engine.Database.removeMeta(Database.java:945)
	at org.h2.table.Table.removeChildrenAndResources(Table.java:553)
	at org.h2.table.TableView.removeChildrenAndResources(TableView.java:425)
	at org.h2.engine.Session.cleanTempTables(Session.java:981)
	at org.h2.engine.Session.close(Session.java:855)
	at org.h2.jdbc.JdbcConnection.close(JdbcConnection.java:405)
	... 3 more

stumc replied:
Latest travis build now reflects error I was seeing. Will revert these fixes back in, whne I have finished the ColumnName class... not long now...

stumc committed:
Demonstration of two meta lock issues

stumc committed:
Added default naming rules

stumc committed:
Configurable column naming policy

stumc committed:
Added help doc

stumc committed:
Created instance class for configuration preventing static Tomcat test case failures

grandinj commented on h2/src/docsrc/help/help.csv:
```diff
@@ -1168,6 +1168,34 @@ This setting can be appended to the database URL: ""jdbc:h2:test;CACHE_SIZE=8192
+SET COLUMN_NAME_RULES=DEFAULT_COLUMN_NAME_PATTERN = '_unnamed$$'; -- Custom configure default column name, $$ is replaced by column Nth ordinal
+SET COLUMN_NAME_RULES=GENERATE_UNIQUE_COLUMN_NAMES = 1; -- Generate unique column names (default is 0, meaning false)
+"
+
```
hmmm, I'm not a big fan of unnecessary flexibility. I'd rather just hardcode a couple of rules for the different database styles we support

but maybe there is some use case I am missing?

stumc replied:
h2 does a really bad job at naming columns with any rules or standards (in fact it makes no effort to do that whatsoever, other than basic SQL alias and SQL expression capture) . I'm used to seeing pretty, generated column names from Oracle and SQL Server. This bug fix scernario included a column name and more then 30 charactes column name - with a CR/LF in it - so I'm trying to regain that nirvana of other DBs. I don't know much about other DB's but they have more rules and capabilities - and that's what I'm trying to capture. 

Whats missing is toUpper() or toLower() support - and column name clash detection (I think Oracle enforces unique names or errors otherwise). If anything there's more to do here than I have had time to do.

grandinj commented on h2/src/test/org/h2/test/scripts/TestScript.java:
```diff
@@ -73,6 +75,10 @@ public static void main(String... a) throws Exception {
+        Path currentRelativePath = Paths.get("");
+        String currentAbsolutePath = currentRelativePath.toAbsolutePath().toString();
+        System.out.println("Current absolute path is: " + currentAbsolutePath);
+        
```
debugging leftovers?

stumc replied:
This test class, annoyingly, wont run outside the TestAll scenario in Eclipse. I want to test my tests standalone outside of TestAll. Whats the trick ? I couldn't get it to work...

grandinj reviewed (changes requested):
looking good

stumc committed:
Removed debugging from TestScript

stumc committed:
Missed some doc items

stumc committed:
Missed some doc items

stumc committed:
Merge branch 'Issue#628' of https://github.com/stumc/h2database into Issue#628

grandinj commented:
I approve of the work in general, but I'd prefer to bind the other-db-emulation-specific rules to the exist MODE=XXX infrastructure. So the ColumnNamer should just look in the Session, find the Mode, and apply the correct rules.

I don't like exposing functionality that I will then have to support later on, unless that functionality really needs to be exposed (I'm willing to be convinced)

stumc committed:
Final round of changes from review

stumc commented:
@grandinj I removed help docs - but the SET commands are still there for the time being, mainly for testing. 

I also integrated it with the Mode setup and added a ModeEnum.

stumc commented:
I found two reasons to keep the the max length identifier setting outside the mode functionality:

1 postgres can be reconfigured to enlarge setting from  60 something default length to 128 with a start up parameter change

2 Oracle latest version sports 128 length so that's what i implemented, but historical versions support only 30 ish. Since the mode does not distinguish versions , users might want to refine manually
